In [1]:
import pandas as pd
import numpy as np


In [50]:
# Importing the dataset
liver_data1=pd.read_csv(r'IndianLiverPatientData.txt',sep="\t",index_col=0,header=None)

In [51]:
liver_data1.head()

1       2     3    4    5   6    7    8    9     10  11
0                                                         
1  65  Female   0.7  0.1  187  16   18  6.8  3.3  0.90  No
2  62    Male  10.9  5.5  699  64  100  7.5  3.2  0.74  No
3  62    Male   7.3  4.1  490  60   68  7.0  3.3  0.89  No
4  58    Male   1.0  0.4  182  14   20  6.8  3.4  1.00  No
5  72    Male   3.9  2.0  195  27   59  7.3  2.4  0.40  No

In [52]:
# giving the variable name
liver_data1.columns=['Age','Gender','Total_Bilirubin','Direct_Bilirubin','Alkaline_Phosphotase',
'Alamine_Aminotransferase','Aspartate_Aminotransferase','Total_Protiens','Albumin',
'Albumin_and_Globulin_Ratio','Class']


In [53]:
liver_data1.head() # crossckecking it by head

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
0                                                                         
1   65  Female              0.7               0.1                   187   
2   62    Male             10.9               5.5                   699   
3   62    Male              7.3               4.1                   490   
4   58    Male              1.0               0.4                   182   
5   72    Male              3.9               2.0                   195   

   Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
0                                                                         
1                        16                          18             6.8   
2                        64                         100             7.5   
3                        60                          68             7.0   
4                        14                          20             6.8   
5                        27                          59             7.3   

   Albumin  Albumin_and_Globulin_Ratio Class  
0                                             
1      3.3                        0.90    No  
2      3.2                        0.74    No  
3      3.3                        0.89    No  
4      3.4                        1.00    No  
5      2.4                        0.40    No

In [54]:
liver_data1.shape # checking the dimension

(583, 11)

In [55]:
liver_data1.info() # To check the data types

<class 'pandas.core.frame.DataFrame'>
Int64Index: 583 entries, 1 to 583
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         583 non-null    int64  
 1   Gender                      563 non-null    object 
 2   Total_Bilirubin             583 non-null    float64
 3   Direct_Bilirubin            583 non-null    float64
 4   Alkaline_Phosphotase        583 non-null    int64  
 5   Alamine_Aminotransferase    583 non-null    int64  
 6   Aspartate_Aminotransferase  583 non-null    int64  
 7   Total_Protiens              568 non-null    float64
 8   Albumin                     583 non-null    float64
 9   Albumin_and_Globulin_Ratio  579 non-null    float64
 10  Class                       583 non-null    object 
dtypes: float64(5), int64(4), object(2)
memory usage: 54.7+ KB


In [56]:
liver_data1.describe()

Age  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
count  583.000000       583.000000        583.000000            583.000000   
mean    44.746141         3.298799          1.486106            290.576329   
std     16.189833         6.209522          2.808498            242.937989   
min      4.000000         0.400000          0.100000             63.000000   
25%     33.000000         0.800000          0.200000            175.500000   
50%     45.000000         1.000000          0.300000            208.000000   
75%     58.000000         2.600000          1.300000            298.000000   
max     90.000000        75.000000         19.700000           2110.000000   

       Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
count                583.000000                  583.000000      568.000000   
mean                  80.713551                  109.910806        6.483979   
std                  182.620356                  288.918529        1.084039   
min                   10.000000                   10.000000        2.700000   
25%                   23.000000                   25.000000        5.800000   
50%                   35.000000                   42.000000        6.600000   
75%                   60.500000                   87.000000        7.200000   
max                 2000.000000                 4929.000000        9.600000   

          Albumin  Albumin_and_Globulin_Ratio  
count  583.000000                  579.000000  
mean     3.141852                    0.947064  
std      0.795519                    0.319592  
min      0.900000                    0.300000  
25%      2.600000                    0.700000  
50%      3.100000                    0.930000  
75%      3.800000                    1.100000  
max      5.500000                    2.800000

In [58]:
liver_data1.isnull().sum() # To check the null vslues

Age                            0
Gender                        20
Total_Bilirubin                0
Direct_Bilirubin               0
Alkaline_Phosphotase           0
Alamine_Aminotransferase       0
Aspartate_Aminotransferase     0
Total_Protiens                15
Albumin                        0
Albumin_and_Globulin_Ratio     4
Class                          0
dtype: int64

In [60]:
print(liver_data1.Gender.value_counts())
liver_data1.Gender.mode()

Male      421
Female    142
Name: Gender, dtype: int64


0    Male
dtype: object

In [61]:
# So replacing the gender variable as Mode is male and it's a categorical variable
liver_data1["Gender"].fillna(liver_data1["Gender"].mode()[0], inplace=True)

In [62]:
# Crosschkeing
liver_data1.isnull().sum()

Age                            0
Gender                         0
Total_Bilirubin                0
Direct_Bilirubin               0
Alkaline_Phosphotase           0
Alamine_Aminotransferase       0
Aspartate_Aminotransferase     0
Total_Protiens                15
Albumin                        0
Albumin_and_Globulin_Ratio     4
Class                          0
dtype: int64

In [63]:
# As total protiens is a numerical variable so filling by median
liver_data1["Total_Protiens"].fillna(liver_data1['Total_Protiens'].median(), inplace = True)

In [64]:
liver_data1.Total_Protiens.isnull().sum() # CRosscheck

0

In [65]:
# As Albumin_and_Globulin_Ratio is a numerical variable so filling by median
liver_data1["Albumin_and_Globulin_Ratio"].fillna(liver_data1['Albumin_and_Globulin_Ratio'].median(), inplace = True)

In [66]:
# Crosscheck
liver_data1.Albumin_and_Globulin_Ratio.isnull().sum()

0

In [67]:
# crosscheck
liver_data1.isnull().sum()

Age                           0
Gender                        0
Total_Bilirubin               0
Direct_Bilirubin              0
Alkaline_Phosphotase          0
Alamine_Aminotransferase      0
Aspartate_Aminotransferase    0
Total_Protiens                0
Albumin                       0
Albumin_and_Globulin_Ratio    0
Class                         0
dtype: int64

In [68]:
colname = ['Gender','Class'] # taking the obeject variable to perform label encoding
colname

['Gender', 'Class']

In [71]:
# Preprocessing the data
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
for x in colname:
    liver_data1[x]  = le.fit_transform(liver_data1[x])

In [72]:
# Crosscheck
liver_data1.head()

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
0                                                                         
1   65       0              0.7               0.1                   187   
2   62       1             10.9               5.5                   699   
3   62       1              7.3               4.1                   490   
4   58       1              1.0               0.4                   182   
5   72       1              3.9               2.0                   195   

   Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
0                                                                         
1                        16                          18             6.8   
2                        64                         100             7.5   
3                        60                          68             7.0   
4                        14                          20             6.8   
5                        27                          59             7.3   

   Albumin  Albumin_and_Globulin_Ratio  Class  
0                                              
1      3.3                        0.90      0  
2      3.2                        0.74      0  
3      3.3                        0.89      0  
4      3.4                        1.00      0  
5      2.4                        0.40      0

In [73]:
# Create x and y
x = liver_data1.values[:,:-1] 
y = liver_data1.values[:,-1] # contain all rows and dependent variable

In [74]:
 # scaling the data
# x and y are created as scaking is only perform on independent variables
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(x)

x = scaler.transform(x)
print(x)

[[ 1.25209764 -1.76228085 -0.41887783 ...  0.29277111  0.19896867
  -0.14752863]
 [ 1.06663704  0.56744644  1.22517135 ...  0.94745542  0.07315659
  -0.65032257]
 [ 1.06663704  0.56744644  0.6449187  ...  0.47982377  0.19896867
  -0.17895325]
 ...
 [ 0.44843504  0.56744644 -0.4027597  ... -0.08133422  0.07315659
   0.16671759]
 [-0.84978917  0.56744644 -0.32216906 ...  0.29277111  0.32478075
   0.16671759]
 [-0.41704777  0.56744644 -0.37052344 ...  0.76040276  1.58290153
   1.73794867]]


In [76]:
liver_data1.dtypes

Age                             int64
Gender                          int32
Total_Bilirubin               float64
Direct_Bilirubin              float64
Alkaline_Phosphotase            int64
Alamine_Aminotransferase        int64
Aspartate_Aminotransferase      int64
Total_Protiens                float64
Albumin                       float64
Albumin_and_Globulin_Ratio    float64
Class                           int32
dtype: object

In [77]:
from sklearn.model_selection import train_test_split

#Spliting the data into test and train
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3,
                                                    random_state=10)  


# Logistic Regression

In [81]:
# Import 
from sklearn.linear_model import LogisticRegression

#Create the model
classifier = LogisticRegression()

#Fitting training data on the model
classifier.fit(x_train,y_train)

#Predicting on the xtest
y_pred = classifier.predict(x_test)
#print(list(zip(y_test,y_pred)))

print(classifier.coef_)
print(classifier.intercept_)

[[-0.41081627 -0.00559557 -0.33993015 -0.86205769 -0.3775615  -1.28541457
  -0.8272897  -0.57187627  0.68971195 -0.2039697 ]]
[-1.72262376]


In [82]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(y_test,y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)



[[114  12]
 [ 40   9]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.74      0.90      0.81       126
         1.0       0.43      0.18      0.26        49

    accuracy                           0.70       175
   macro avg       0.58      0.54      0.54       175
weighted avg       0.65      0.70      0.66       175

Accuracy of the model:  0.7028571428571428


In [84]:
# Stored the predicted probabilities

y_pred_prob = classifier.predict_proba(x_test)
y_pred_prob

array([[7.55018206e-01, 2.44981794e-01],
       [9.88634452e-01, 1.13655480e-02],
       [3.96343302e-01, 6.03656698e-01],
       [8.37244213e-01, 1.62755787e-01],
       [6.34793656e-01, 3.65206344e-01],
       [5.34543172e-01, 4.65456828e-01],
       [5.39965234e-01, 4.60034766e-01],
       [9.09399886e-01, 9.06001140e-02],
       [6.44045278e-01, 3.55954722e-01],
       [9.99999746e-01, 2.54295272e-07],
       [9.97542316e-01, 2.45768404e-03],
       [4.20288354e-01, 5.79711646e-01],
       [7.46831670e-01, 2.53168330e-01],
       [6.53345995e-01, 3.46654005e-01],
       [8.92459662e-01, 1.07540338e-01],
       [9.94286795e-01, 5.71320536e-03],
       [3.95071768e-01, 6.04928232e-01],
       [7.19008255e-01, 2.80991745e-01],
       [9.17885440e-01, 8.21145596e-02],
       [7.61324002e-01, 2.38675998e-01],
       [5.02912907e-01, 4.97087093e-01],
       [7.59862427e-01, 2.40137573e-01],
       [7.62214720e-01, 2.37785280e-01],
       [3.79428759e-01, 6.20571241e-01],
       [7.307851

In [87]:
for a in np.arange(0.4,0.61,0.01):
    predict_mine = np.where(y_pred_prob[:,1] > a, 1, 0)
    cfm=confusion_matrix(y_test, predict_mine)
    total_err=cfm[0,1]+cfm[1,0]
    print("Errors at threshold ", a, ":",total_err, " , type 2 error :", 
          cfm[1,0]," , type 1 error:", cfm[0,1])

Errors at threshold  0.4 : 58  , type 2 error : 33  , type 1 error: 25
Errors at threshold  0.41000000000000003 : 55  , type 2 error : 33  , type 1 error: 22
Errors at threshold  0.42000000000000004 : 55  , type 2 error : 35  , type 1 error: 20
Errors at threshold  0.43000000000000005 : 55  , type 2 error : 35  , type 1 error: 20
Errors at threshold  0.44000000000000006 : 55  , type 2 error : 36  , type 1 error: 19
Errors at threshold  0.45000000000000007 : 56  , type 2 error : 37  , type 1 error: 19
Errors at threshold  0.4600000000000001 : 55  , type 2 error : 37  , type 1 error: 18
Errors at threshold  0.4700000000000001 : 53  , type 2 error : 38  , type 1 error: 15
Errors at threshold  0.4800000000000001 : 51  , type 2 error : 38  , type 1 error: 13
Errors at threshold  0.4900000000000001 : 50  , type 2 error : 38  , type 1 error: 12
Errors at threshold  0.5000000000000001 : 52  , type 2 error : 40  , type 1 error: 12
Errors at threshold  0.5100000000000001 : 51  , type 2 error : 4

In [88]:
y_pred_class = []
for value in y_pred_prob[:,1]:
    if value > 0.57:
        y_pred_class.append(1)
    else:
        y_pred_class.append(0)
print(y_pred_class)

[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


In [90]:

from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
cfm=confusion_matrix(y_test,y_pred_class)
print(cfm)
acc=accuracy_score(y_test, y_pred_class)
print("Accuracy of the model: ",acc)
print(classification_report(y_test, y_pred_class))


[[123   3]
 [ 43   6]]
Accuracy of the model:  0.7371428571428571
              precision    recall  f1-score   support

         0.0       0.74      0.98      0.84       126
         1.0       0.67      0.12      0.21        49

    accuracy                           0.74       175
   macro avg       0.70      0.55      0.52       175
weighted avg       0.72      0.74      0.66       175



## Conclusion(Logistic Regression)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and new to be get diagnosed in future
### Here doing logistic regression on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.737) but the recall for predicting 1(it's a patient) is predicted very low which i.e.(0.12) even 
#### overall accuracy 1 is 0.21 which is very low 
### so, this algorithm is not performing good in this dataset

# Knn

In [94]:
#predicting using the KNeighbors_Classifier
from sklearn.neighbors import KNeighborsClassifier
model_KNN=KNeighborsClassifier(n_neighbors=int(np.sqrt(len(x_train))), metric='euclidean')
#fit the model on the data and predict the values
model_KNN.fit(x_train,y_train)

Y_pred=model_KNN.predict(x_test)
#print(list(zip(Y_test,Y_pred)))

In [95]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(y_test,y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)



[[114  12]
 [ 40   9]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.74      0.90      0.81       126
         1.0       0.43      0.18      0.26        49

    accuracy                           0.70       175
   macro avg       0.58      0.54      0.54       175
weighted avg       0.65      0.70      0.66       175

Accuracy of the model:  0.7028571428571428


In [96]:
int(np.sqrt(len(x_train)))

20

In [99]:
# Selecting best k values
## selecting best value for k
from sklearn.metrics import accuracy_score
my_dict={}
for K in range(1,21):
    model_KNN = KNeighborsClassifier(n_neighbors=K,metric="euclidean")
    model_KNN.fit(x_train, y_train)
    y_pred = model_KNN.predict(x_test)
    print ("Accuracy is ", accuracy_score(y_test,Y_pred), "for K-Value:",K)
    my_dict[K]=accuracy_score(y_test,y_pred)

Accuracy is  0.6914285714285714 for K-Value: 1
Accuracy is  0.6914285714285714 for K-Value: 2
Accuracy is  0.6914285714285714 for K-Value: 3
Accuracy is  0.6914285714285714 for K-Value: 4
Accuracy is  0.6914285714285714 for K-Value: 5
Accuracy is  0.6914285714285714 for K-Value: 6
Accuracy is  0.6914285714285714 for K-Value: 7
Accuracy is  0.6914285714285714 for K-Value: 8
Accuracy is  0.6914285714285714 for K-Value: 9
Accuracy is  0.6914285714285714 for K-Value: 10
Accuracy is  0.6914285714285714 for K-Value: 11
Accuracy is  0.6914285714285714 for K-Value: 12
Accuracy is  0.6914285714285714 for K-Value: 13
Accuracy is  0.6914285714285714 for K-Value: 14
Accuracy is  0.6914285714285714 for K-Value: 15
Accuracy is  0.6914285714285714 for K-Value: 16
Accuracy is  0.6914285714285714 for K-Value: 17
Accuracy is  0.6914285714285714 for K-Value: 18
Accuracy is  0.6914285714285714 for K-Value: 19
Accuracy is  0.6914285714285714 for K-Value: 20


In [136]:
## Conclusion(Knn)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and neww to be get diagnosed in future
### Here doing Knn on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.702) but the recall for predicting 1(it's a patient) is predicted very low which i.e.(0.18) even 
#### overall accuracy 1 is 0.26 which is very low 
### so, this algorithm is not performing good in this dataset

# SVM

In [100]:
from sklearn import svm
svc_model=svm.SVC(kernel='rbf',C=1.0,gamma=0.1)
svc_model.fit(x_train, y_train)
y_pred=svc_model.predict(x_test)
#print(list(Y_pred))


In [101]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(y_test,y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)

[[122   4]
 [ 46   3]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.73      0.97      0.83       126
         1.0       0.43      0.06      0.11        49

    accuracy                           0.71       175
   macro avg       0.58      0.51      0.47       175
weighted avg       0.64      0.71      0.63       175

Accuracy of the model:  0.7142857142857143


In [106]:
#using cross validation
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier

classifier = svm.SVC(kernel='rbf', C= 1.0, gamma=0.1) # 71.30%
#classifier = KNeighborsClassifier(n_neighbors=20, metric='euclidean') #71.07%
#classifier = svm.SVC(kernel='rbf',C=10.0, gamma=0.001) # 71.05%
#classifier = LogisticRegression() # 73.25%

# Performing kfold_cross_validation
from sklearn.model_selection import KFold
kfold_cv = KFold(n_splits=10)
print(kfold_cv)

from sklearn.model_selection import cross_val_score

# Running the model using scoring metric as acuracy
kfold_cv_result = cross_val_score(estimator= classifier, X= x_train,
                                 y= y_train, cv = kfold_cv)

print(kfold_cv_result)

# Finding the mean
print(kfold_cv_result.mean())

KFold(n_splits=10, random_state=None, shuffle=False)
[0.82926829 0.58536585 0.75609756 0.53658537 0.73170732 0.73170732
 0.73170732 0.87804878 0.625      0.725     ]
0.7130487804878048


In [ ]:
## Conclusion(svm)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and neww to be get diagnosed in future
### Here doing svm on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.714) but the recall for predicting 1(it's a patient) is predicted very low which i.e.(0.06) even 
#### overall accuracy 1 is 0.11 which is very low 
### so, this algorithm is not performing good in this dataset

# Decision Tree

In [109]:
#predicting using the Decision_Tree_Classifier
from sklearn.tree import DecisionTreeClassifier

model_DecisionTree=DecisionTreeClassifier(random_state=10)

#fit the model on the data and predict the values
model_DecisionTree.fit(x_train,y_train)
y_pred=model_DecisionTree.predict(x_test)

In [110]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(y_test,y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)


[[99 27]
 [25 24]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.80      0.79      0.79       126
         1.0       0.47      0.49      0.48        49

    accuracy                           0.70       175
   macro avg       0.63      0.64      0.64       175
weighted avg       0.71      0.70      0.70       175

Accuracy of the model:  0.7028571428571428


# Random Forest

In [111]:
#predicting using the Random_Forest_Classifier
from sklearn.ensemble import RandomForestClassifier

model_RandomForest=RandomForestClassifier(n_estimators=100,random_state=10)

#fit the model on the data and predict the values
model_RandomForest.fit(x_train,y_train)

y_pred=model_RandomForest.predict(x_test)


In [112]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(y_test,y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)


[[108  18]
 [ 34  15]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.76      0.86      0.81       126
         1.0       0.45      0.31      0.37        49

    accuracy                           0.70       175
   macro avg       0.61      0.58      0.59       175
weighted avg       0.67      0.70      0.68       175

Accuracy of the model:  0.7028571428571428


# Upsampling

In [113]:
liver_data1.Class.value_counts()

0    416
1    167
Name: Class, dtype: int64

In [115]:
from sklearn.utils import resample
# Separate majority and minority classes
df_majority = liver_data1[liver_data1.Class==0]
df_minority = liver_data1[liver_data1.Class==1]
 
# Upsample minority class
df_minority_upsampled = resample(df_minority, 
                                 replace=True,     # sample with replacement
                                 n_samples=416,    # to match majority class
                                 random_state=10) # reproducible results
 
# Combine majority class with upsampled minority class
df_upsampled = pd.concat([df_majority, df_minority_upsampled])
 
# Display new class counts
df_upsampled.Class.value_counts()

1    416
0    416
Name: Class, dtype: int64

In [116]:
X = df_upsampled.values[:,:-1]
Y = df_upsampled.values[:,-1]

In [118]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(X)

X = scaler.transform(X)
print(X)

[[ 1.31929811 -1.73762012 -0.36813594 ...  0.25720402  0.12505717
  -0.20405777]
 [ 1.13104244  0.57549978  1.55193588 ...  0.90848934  0.00192396
  -0.7137045 ]
 [ 1.13104244  0.57549978  0.87426347 ...  0.44328554  0.12505717
  -0.23591069]
 ...
 [-0.18674721 -1.73762012 -0.33048747 ...  1.0015301   0.74072321
   0.11447143]
 [-0.87701797 -1.73762012 -0.34931171 ...  1.28065238  1.60265568
   1.07005905]
 [-0.56325853  0.57549978 -0.33048747 ... -0.11495902  0.4944568
   0.75152984]]


In [119]:
from sklearn.model_selection import train_test_split

#Split the data into test and train
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3,
                                                    random_state=10)  


In [120]:
#predicting using the Decision_Tree_Classifier
from sklearn.tree import DecisionTreeClassifier

model_DecisionTree=DecisionTreeClassifier(random_state=10)

#fit the model on the data and predict the values
model_DecisionTree.fit(X_train,Y_train)
Y_pred=model_DecisionTree.predict(X_test)

In [121]:
from sklearn.metrics import confusion_matrix, accuracy_score,\
classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)


[[ 93  37]
 [  8 112]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.92      0.72      0.81       130
         1.0       0.75      0.93      0.83       120

    accuracy                           0.82       250
   macro avg       0.84      0.82      0.82       250
weighted avg       0.84      0.82      0.82       250

Accuracy of the model:  0.82


In [137]:
## Conclusion(Decision Tree)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and neww to be get diagnosed in future
### Here doing Decision Tree on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.82) but the recall for predicting 1(it's a patient) is predicted 
#### good in this case which i.e.(0.93) even overall accuracy 1 is 0.83 which is very low 
### so, this algorithm is performing good in this dataset

In [122]:
#predicting using the Random_Forest_Classifier
from sklearn.ensemble import RandomForestClassifier

model_RandomForest=RandomForestClassifier(n_estimators=100,random_state=10)

#fit the model on the data and predict the values
model_RandomForest.fit(X_train,Y_train)

Y_pred=model_RandomForest.predict(X_test)


In [123]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)


[[ 93  37]
 [  7 113]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.93      0.72      0.81       130
         1.0       0.75      0.94      0.84       120

    accuracy                           0.82       250
   macro avg       0.84      0.83      0.82       250
weighted avg       0.85      0.82      0.82       250

Accuracy of the model:  0.824


In [138]:
## Conclusion(Random Forest)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and neww to be get diagnosed in future
### Here doing Random forest on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.824) but the recall for predicting 1(it's a patient) is predicted 
#### good in this case which i.e.(0.94) even overall accuracy 1 is 0.84 which is very low 
### so, this algorithm is performing good in this dataset

# ANN-MLP Classifier

In [126]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(12,12,12),max_iter=100, early_stopping=True,n_iter_no_change=5,
                    random_state=10, activation="relu", solver="adam",learning_rate="constant",
                    learning_rate_init=0.01,verbose=True)
mlp.fit(x_train,y_train)
y_pred = mlp.predict(x_test)
#print(list(zip(Y_test,Y_pred)))

Iteration 1, loss = 0.59326202
Validation score: 0.707317
Iteration 2, loss = 0.57486243
Validation score: 0.707317
Iteration 3, loss = 0.56442650
Validation score: 0.707317
Iteration 4, loss = 0.55331315
Validation score: 0.707317
Iteration 5, loss = 0.54242179
Validation score: 0.707317
Iteration 6, loss = 0.53602994
Validation score: 0.707317
Iteration 7, loss = 0.52995012
Validation score: 0.707317
Validation score did not improve more than tol=0.000100 for 5 consecutive epochs. Stopping.


In [127]:
from sklearn.metrics import confusion_matrix, accuracy_score,\
classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)



[[ 71  59]
 [ 13 107]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.72      1.00      0.84       126
         1.0       0.00      0.00      0.00        49

    accuracy                           0.72       175
   macro avg       0.36      0.50      0.42       175
weighted avg       0.52      0.72      0.60       175

Accuracy of the model:  0.72


C:\Users\Soni Mrunali\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
## Conclusion(ANN)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and neww to be get diagnosed in future
### Here doing ANN on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.72) but the recall for predicting 1(it's a patient) is predicted very low which i.e.(0.00) even 
#### overall accuracy 1 is 0.00 which is very low 
### so, this algorithm is not performing good in this dataset

# Grid Searchcv

In [128]:
from sklearn.neural_network import MLPClassifier

mlp_gs = MLPClassifier(max_iter=100,early_stopping=True,n_iter_no_change=5,
                    random_state=10,learning_rate_init=0.01, learning_rate="constant",
                   verbose=True )
parameter_space = {
    'hidden_layer_sizes': [(12,12,12),(10,10),(100)],
    'activation': ['tanh', 'relu'],
    'solver': ['sgd', 'adam'],
    'alpha': [0.0001, 0.01]
}
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(mlp_gs, parameter_space, n_jobs=-1, cv=5)


In [131]:
clf.fit(x_train,y_train)

Iteration 1, loss = 0.65354498
Validation score: 0.682927
Iteration 2, loss = 0.55076909
Validation score: 0.682927
Iteration 3, loss = 0.52469121
Validation score: 0.682927
Iteration 4, loss = 0.51842560
Validation score: 0.707317
Iteration 5, loss = 0.51054754
Validation score: 0.707317
Iteration 6, loss = 0.50356338
Validation score: 0.756098
Iteration 7, loss = 0.49860141
Validation score: 0.756098
Iteration 8, loss = 0.49551149
Validation score: 0.756098
Iteration 9, loss = 0.49206033
Validation score: 0.780488
Iteration 10, loss = 0.48935834
Validation score: 0.780488
Iteration 11, loss = 0.48623440
Validation score: 0.780488
Iteration 12, loss = 0.48082325
Validation score: 0.780488
Iteration 13, loss = 0.47860527
Validation score: 0.707317
Iteration 14, loss = 0.47568270
Validation score: 0.707317
Iteration 15, loss = 0.47340923
Validation score: 0.707317
Validation score did not improve more than tol=0.000100 for 5 consecutive epochs. Stopping.


GridSearchCV(cv=5,
             estimator=MLPClassifier(early_stopping=True,
                                     learning_rate_init=0.01, max_iter=100,
                                     n_iter_no_change=5, random_state=10,
                                     verbose=True),
             n_jobs=-1,
             param_grid={'activation': ['tanh', 'relu'],
                         'alpha': [0.0001, 0.01],
                         'hidden_layer_sizes': [(12, 12, 12), (10, 10), 100],
                         'solver': ['sgd', 'adam']})

In [132]:
print('Best parameters found:\n', clf.best_params_)

Best parameters found:
 {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': 100, 'solver': 'adam'}


In [134]:
Y_pred = clf.predict(x_test)
print(list(zip(y_test,y_pred)))

[(0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (1.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (1.0, 0.0), (0.0, 0.0), (1.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (0.0, 0.0), (1.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (1.0, 0.0), (1.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.0, 0.0), (0.

In [135]:
from sklearn.metrics import confusion_matrix, accuracy_score,\
classification_report

cfm=confusion_matrix(y_test,y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(y_test,y_pred))

acc=accuracy_score(y_test, y_pred)
print("Accuracy of the model: ",acc)

[[126   0]
 [ 49   0]]
Classification report: 
              precision    recall  f1-score   support

         0.0       0.72      1.00      0.84       126
         1.0       0.00      0.00      0.00        49

    accuracy                           0.72       175
   macro avg       0.36      0.50      0.42       175
weighted avg       0.52      0.72      0.60       175

Accuracy of the model:  0.72


C:\Users\Soni Mrunali\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
## Conclusion(Greed searchcv)
### It contains 2 data values and they are:
#### 0 mean it's not a patient and does not need diagnosed in future
#### 1 mean it's a patient and neww to be get diagnosed in future
### Here doing Greedsearch on training data and predicted the values on test data.
#### In this it has shown good overall accuracy(0.824) but the recall for predicting 1(it's a patient) is not predicted 
#### good in this case which i.e.(0.00) even overall accuracy 1 is 0.84 which is very low 
### so, this algorithm is not performing good in this dataset

# Conclusion
## So, I can conclude here that Random forest with(0.824) and Decision Tree with(0.82)
### The data was supervised and prediction was binary that's why i have choosen classification models